lib/Adapter/WorseReflection/Helper/WorseUnresolvableClassNameFinder.php

```diff
@@ -122,6 +122,12 @@ private function appendUnresolvedName(QualifiedName $name, array $unresolvedName
 
     private function appendUnresolvedClassName(string $nameText, array $unresolvedNames, QualifiedName $name): array
     {
+        if (current(array_filter($unresolvedNames, function (NameWithByteOffset $name) use ($nameText) {
+            return (string)$name->name() === $nameText;
+        })) !== false) {
+            return $unresolvedNames;
+        }
```

Comment:
Should probably add a benchmark for this. You are checking here if the unresolved name already exists? The byte offset is important here, so this might not be correct.

Reply:
Yes, I am checking if the name has been reported. My assumption is that for a given `$nameText` (which includes the namespace the current file is in) in a single file there will be a fixed number of suggestions regardless of the offset. I can't think of why that would be wrong.

Reply:
I mean, even if we return the name multiple times because of the offset (VS Code doesn't care about this, but VIM probably does), then at least we should skip tyring to resolve it. Furthermore, not sure where suggestions are calculated, but these should not be calculated multiple times either. That thing is really slow if you paste a whole class with lots of code.

Reply:
It's a separte instance of the name, it's used for diagnostics.

Reply:
Good, do one for a whole file, say 20-30 suggestions.
So the deduping should be done in the code actions in LSP?

Reply:
See: https://github.com/phpactor/code-transform/pull/55

Reply:
So with this code compared against master there is no significant difference:

```

+--------------------------+----------------+--------------------+
| set                      | mem_peak       | mode               |
+--------------------------+----------------+--------------------+
| class/empty              | 5.236mb +1.38% | 77.117μs +3.84%    |
| class/missing1           | 5.407mb +0.03% | 273.151μs +0.70%   |
| class/missing10different | 5.796mb +0.04% | 1,810.896μs +5.93% |
| class/missing10same      | 5.838mb +0.04% | 1,781.656μs -5.48% |
| func/empty               | 5.236mb +1.38% | 77.587μs +3.52%    |
| func/missing1            | 5.398mb +0.02% | 301.996μs +0.46%   |
| func/missing10different  | 5.718mb +0.04% | 2,505.979μs +1.05% |
| func/missing10same       | 5.718mb +0.04% | 2,487.186μs +0.54% |
+--------------------------+----------------+--------------------+
```

Reply:
It's possibe that WR isn't caching NOT FOUND however (becuase it's an exception), which is why the time is the same for 10 same and 10 different probably.

Reply:
Is that Benchmark affected by the scale of the project? I.e. large projects running slower?

The performance problem should be [here](https://github.com/phpactor/language-server-phpactor-extensions/blob/cad1ecefe3253cfaae6ce12a5513b94d477d4e9b/lib/LanguageServerCodeTransform/Model/NameImport/CandidateFinder.php#L66-L75) where the candidates are being calculated. Right?

Reply:
I added an artificial delay to the source location  (which was instant initially) and fixed the caching in WR to cache exceptions:

```
+--------------------------+----------------+---------------------+
| set                      | mem_peak       | mode                |
+--------------------------+----------------+---------------------+
| class/empty              | 5.306mb +1.17% | 77.166μs -0.43%     |
| class/missing1           | 5.406mb -0.13% | 269.351μs -19.56%   |
| class/missing10different | 5.796mb +1.23% | 1,697.144μs -27.72% |
| class/missing10same      | 5.838mb +0.40% | 1,849.910μs -36.70% |
| func/empty               | 5.306mb +1.17% | 77.502μs -0.13%     |
| func/missing1            | 5.399mb -0.29% | 299.722μs -17.71%   |
| func/missing10different  | 5.718mb +1.18% | 2,476.679μs -20.86% |
| func/missing10same       | 5.718mb +0.35% | 2,436.735μs -22.54% |
+--------------------------+----------------+---------------------+
```

Reply:
So that should provide better performance in general (and the benefit increases the more things were not found)

Reply:
But still we should not calculate candidates multiple times. I will PR that.

Reply:
I keep getting:
```
[Error - 5:32:38 PM] Request textDocument/codeAction failed.
  Message: Exception [TypeError] Return value of Phpactor\WorseReflection\Core\Reflector\ClassReflector\MemonizedReflector::sourceCodeForFunction() must be an instance of Phpactor\WorseReflection\Core\SourceCode, bool returned
  Code: -32603 
#0 /root/phpactor/worse-reflection/lib/Core/Reflector/CompositeReflector.php(127): Phpactor\WorseReflection\Core\Reflector\ClassReflector\MemonizedReflector->sourceCodeForFunction()
#1 /root/phpactor/code-transform/lib/Adapter/WorseReflection/Helper/WorseUnresolvableClassNameFinder.php(154): Phpactor\WorseReflection\Core\Reflector\CompositeReflector->sourceCodeForFunction()
#2 /root/phpactor/code-transform/lib/Adapter/WorseReflection/Helper/WorseUnresolvableClassNameFinder.php(84): Phpactor\CodeTransform\Adapter\WorseReflection\Helper\WorseUnresolvableClassNameFinder->appendUnresolvedFunctionName()
#3 /root/phpactor/code-transform/lib/Adapter/WorseReflection/Helper/WorseUnresolvableClassNameFinder.php(70): Phpactor\CodeTransform\Adapter\WorseReflection\Helper\WorseUnresolvableClassNameFinder->appendUnresolvedName()
#4 /root/phpactor/code-transform/lib/Adapter/WorseReflection/Helper/WorseUnresolvableClassNameFinder.php(50): Phpactor\CodeTransform\Adapter\WorseReflection\Helper\WorseUnresolvableClassNameFinder->filterResolvedNames()
#5 /root/phpactor/language-server-phpactor-extensions/lib/LanguageServerCodeTransform/Model/NameImport/CandidateFinder.php(51): Phpactor\CodeTransform\Adapter\WorseReflection\Helper\WorseUnresolvableClassNameFinder->find()
#6 /root/phpactor/language-server-phpactor-extensions/lib/LanguageServerCodeTransform/Model/NameImport/CandidateFinder.php(70): Phpactor\Extension\LanguageServerCodeTransform\Model\NameImport\CandidateFinder->unresolved()
#7 /root/phpactor/language-server-phpactor-extensions/lib/LanguageServerCodeTransform/CodeAction/ImportNameProvider.php(39): Phpactor\Extension\LanguageServerCodeTransform\Model\NameImport\CandidateFinder->importCandidates()
#8 [internal function]: Phpactor\Extension\LanguageServerCodeTransform\CodeAction\ImportNameProvider->Phpactor\Extension\LanguageServerCodeTransform\CodeAction\{closure}()
#9 /root/phpactor/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#10 /root/phpactor/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#11 /root/phpactor/language-server-phpactor-extensions/lib/LanguageServerCodeTransform/CodeAction/ImportNameProvider.php(49): Amp\call()
#12 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/CodeAction/AggregateCodeActionProvider.php(30): Phpactor\Extension\LanguageServerCodeTransform\CodeAction\ImportNameProvider->provideActionsFor()
#13 [internal function]: Phpactor\LanguageServer\Core\CodeAction\AggregateCodeActionProvider->Phpactor\LanguageServer\Core\CodeAction\{closure}()
#14 /root/phpactor/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#15 /root/phpactor/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#16 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/CodeAction/AggregateCodeActionProvider.php(34): Amp\call()
#17 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Handler/TextDocument/CodeActionHandler.php(59): Phpactor\LanguageServer\Core\CodeAction\AggregateCodeActionProvider->provideActionsFor()
#18 /root/phpactor/phpactor/vendor/amphp/amp/lib/functions.php(90): Phpactor\LanguageServer\Handler\TextDocument\CodeActionHandler->Phpactor\LanguageServer\Handler\TextDocument\{closure}()
#19 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Handler/TextDocument/CodeActionHandler.php(60): Amp\call()
#20 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Handler/HandlerMethodRunner.php(86): Phpactor\LanguageServer\Handler\TextDocument\CodeActionHandler->codeAction()
#21 [internal function]: Phpactor\LanguageServer\Core\Handler\HandlerMethodRunner->Phpactor\LanguageServer\Core\Handler\{closure}()
#22 /root/phpactor/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#23 /root/phpactor/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#24 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Handler/HandlerMethodRunner.php(102): Amp\call()
#25 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Middleware/HandlerMiddleware.php(38): Phpactor\LanguageServer\Core\Handler\HandlerMethodRunner->dispatch()
#26 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\HandlerMiddleware->process()
#27 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Middleware/ResponseHandlingMiddleware.php(35): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#28 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\ResponseHandlingMiddleware->process()
#29 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Middleware/MethodAliasMiddleware.php(37): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#30 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\MethodAliasMiddleware->process()
#31 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Middleware/CancellationMiddleware.php(49): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#32 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\CancellationMiddleware->process()
#33 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Middleware/InitializeMiddleware.php(68): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#34 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\InitializeMiddleware->process()
#35 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Middleware/ErrorHandlingMiddleware.php(39): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#36 [internal function]: Phpactor\LanguageServer\Middleware\ErrorHandlingMiddleware->Phpactor\LanguageServer\Middleware\{closure}()
#37 /root/phpactor/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#38 /root/phpactor/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#39 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Middleware/ErrorHandlingMiddleware.php(64): Amp\call()
#40 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\ErrorHandlingMiddleware->process()
#41 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Dispatcher/Dispatcher/MiddlewareDispatcher.php(29): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#42 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(223): Phpactor\LanguageServer\Core\Dispatcher\Dispatcher\MiddlewareDispatcher->dispatch()
#43 [internal function]: Phpactor\LanguageServer\Core\Server\LanguageServer->Phpactor\LanguageServer\Core\Server\{closure}()
#44 /root/phpactor/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#45 /root/phpactor/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#46 /root/phpactor/phpactor/vendor/amphp/amp/lib/functions.php(121): Amp\call()
#47 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(240): Amp\asyncCall()
#48 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(214): Phpactor\LanguageServer\Core\Server\LanguageServer->dispatchRequest()
#49 [internal function]: Phpactor\LanguageServer\Core\Server\LanguageServer->Phpactor\LanguageServer\Core\Server\{closure}()
#50 /root/phpactor/phpactor/vendor/amphp/amp/lib/Coroutine.php(118): Generator->send()
#51 /root/phpactor/phpactor/vendor/amphp/amp/lib/Internal/Placeholder.php(149): Amp\Coroutine->Amp\{closure}()
#52 /root/phpactor/phpactor/vendor/amphp/amp/lib/Coroutine.php(123): Amp\Coroutine->resolve()
#53 /root/phpactor/phpactor/vendor/amphp/amp/lib/Internal/Placeholder.php(149): Amp\Coroutine->Amp\{closure}()
#54 /root/phpactor/phpactor/vendor/amphp/amp/lib/Deferred.php(52): class@anonymous->resolve()
#55 /root/phpactor/phpactor/vendor/amphp/byte-stream/lib/ResourceInputStream.php(101): Amp\Deferred->resolve()
#56 /root/phpactor/phpactor/vendor/amphp/amp/lib/Loop/NativeDriver.php(321): Amp\ByteStream\ResourceInputStream::Amp\ByteStream\{closure}()
#57 /root/phpactor/phpactor/vendor/amphp/amp/lib/Loop/NativeDriver.php(127): Amp\Loop\NativeDriver->selectStreams()
#58 /root/phpactor/phpactor/vendor/amphp/amp/lib/Loop/Driver.php(138): Amp\Loop\NativeDriver->dispatch()
#59 /root/phpactor/phpactor/vendor/amphp/amp/lib/Loop/Driver.php(72): Amp\Loop\Driver->tick()
#60 /root/phpactor/phpactor/vendor/amphp/amp/lib/Loop.php(95): Amp\Loop\Driver->run()
#61 /root/phpactor/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(113): Amp\Loop::run()
#62 /root/phpactor/phpactor/vendor/phpactor/language-server-extension/lib/LanguageServer/Command/StartCommand.php(59): Phpactor\LanguageServer\Core\Server\LanguageServer->run()
#63 /root/phpactor/phpactor/vendor/symfony/console/Command/Command.php(256): Phpactor\Extension\LanguageServer\Command\StartCommand->execute()
#64 /root/phpactor/phpactor/vendor/symfony/console/Application.php(971): Symfony\Component\Console\Command\Command->run()
#65 /root/phpactor/phpactor/vendor/symfony/console/Application.php(290): Symfony\Component\Console\Application->doRunCommand()
#66 /root/phpactor/phpactor/lib/Application.php(54): Symfony\Component\Console\Application->doRun()
#67 /root/phpactor/phpactor/vendor/symfony/console/Application.php(166): Phpactor\Application->doRun()
#68 /root/phpactor/phpactor/bin/phpactor(42): Symfony\Component\Console\Application->run()
#69 {main}
```

What am I missing?

Reply:
yes i just fixed that a second ago

Reply:
Right, thanks. I will PR [this then](https://github.com/phpactor/language-server-phpactor-extensions/blob/cad1ecefe3253cfaae6ce12a5513b94d477d4e9b/lib/LanguageServerCodeTransform/Model/NameImport/CandidateFinder.php#L66-L75).